File: yandex_eserajim_sleepless_nights/5. Покрытие K отрезками/solution.cs
using System;
using System.Linq;

class Program 
{
    static void Main() 
    {
        string[] inputLine = Console.ReadLine().Split();
        int n = int.Parse(inputLine[0]);
        int k = int.Parse(inputLine[1]);

        int[] points = Console.ReadLine().Split().Select(int.Parse).OrderBy(x => x).ToArray();

        int left = 0;
        int right = points[n - 1] - points[0];
        int result = right;

        while (left <= right)
        {
            int mid = left + (right - left) / 2; // Избегаем переполнения
            if (CanCover(points, k, mid))
            {
                result = mid;
                right = mid - 1; // Пытаемся найти еще меньший отрезок
            }
            else
            {
                left = mid + 1; // Увеличиваем длину отрезка
            }
        }

        Console.WriteLine(result);
    }

    private static bool CanCover(int[] points, int k, int length)
    {
        int segmentsUsed = 1;
        int currentSegmentStart = points[0];
        
        for(int i = 1; i < points.Length; i++)
        {
            if(points[i] > currentSegmentStart + length)
            {
                segmentsUsed++;
                currentSegmentStart = points[i];
            }
        }

        return segmentsUsed <= k;
    }
}

File: yandex_eserajim_sleepless_nights/5. Покрытие K отрезками/solution_.py
import sys
from typing import List

def can_cover(points: List[int], k: int, length: int) -> bool:
    segments_used: int = 1
    current_segment_start: int = points[0]

    for point in points[1:]:
        if point > current_segment_start + length:
            segments_used += 1
            current_segment_start = point
    return segments_used <= k


def find_min_length(points: List[int], k: int) -> int:
    points.sort()
    left: int = 0
    right: int = points[-1] - points[0]
    result: int = right

    while left <= right:
        mid: int = left + (right - left) // 2
        if can_cover(points, k, mid):
            result = mid
            right = mid - 1
        else:
            left = mid + 1
    return result


def main() -> None:
    n, k = map(int, input().split())
    points = list(map(int, input().split()))
    print(find_min_length(points, k))

if __name__ == "__main__":
    main()